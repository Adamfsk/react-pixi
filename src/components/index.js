import BitmapText from './BitmapText'
import Container from './Container'
import Graphics from './Graphics'
import NineSlicePlane from './NineSlicePlane'
import ParticleContainer from './ParticleContainer'
import Sprite from './Sprite'
import Stage from './Stage'
import Text from './Text'
import TilingSprite from './TilingSprite'

export {
  BitmapText,
  Container,
  Graphics,
  NineSlicePlane,
  ParticleContainer,
  Sprite,
  Stage,
  Text,
  TilingSprite
}
