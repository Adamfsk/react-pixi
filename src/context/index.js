import React from 'react'
import PropTypes from 'prop-types'
import invariant from 'fbjs/lib/invariant'
import idx from 'idx'

let Context

function clearContext() {
  Context = undefined
}

function createContext() {
  if ('createContext' in React && typeof React.createContext === 'function') {
    Context = React.createContext('pixi-app')
  }
  return Context
}

function createProvider() {
  let Provider = Context && Context.Consumer
  if (!Provider) {
    class PixiAppProvider extends React.Component {
      componentWillMount() {
        const app = idx(this, _ => _.context.app)
        invariant(
          !!app,
          'Cannot provide `PIXI.Application` via React Context. You probably forgot to wrap your PIXI application with' +
            ' <Stage /> object'
        )
      }

      render() {
        return this.props.children(this.context.app)
      }
    }

    PixiAppProvider.propTypes = {
      children: PropTypes.func,
    }

    PixiAppProvider.defaultProps = {
      children: () => {},
    }

    PixiAppProvider.contextTypes = {
      app: PropTypes.object,
    }

    Provider = PixiAppProvider
    Provider.displayName = 'PixiAppProvider'
  }

  return Provider
}

export { Context, createProvider, createContext, clearContext }
