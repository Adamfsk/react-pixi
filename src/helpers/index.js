export const isFunction = (...args) => {
  return args.reduce((a, b) => a && typeof b === 'function', true)
}

export const isObject = obj => Object.prototype.toString.call(obj) === '[object Object]'

export const hasKey = collection => {
  let coll = collection

  if (!Array.isArray(coll)) {
    if (isObject(collection)) {
      coll = Object.keys(collection)
    } else {
      throw new Error('collection needs to be an Array or Object')
    }
  }

  const index = {}
  coll.forEach(key => {
    index[key] = true
  })

  return key => typeof index[key] !== 'undefined'
}

// See https://github.com/michalochman/react-pixi-fiber/blob/a4dbddbef0ffbf0f563c64d30766ea28222a51ea/src/utils.js#L7
export const not = boolFn => (...args) => !boolFn(...args)

export const runningInBrowser = () => Boolean(window)
