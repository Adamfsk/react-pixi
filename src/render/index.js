import invariant from 'fbjs/lib/invariant'
import { Container } from 'pixi.js'
import PixiFiber, { PACKAGE_NAME, VERSION } from '../reconciler'

// cache root containers
export const roots = new Map()

/**
 * Renderer
 *
 * @param {any} element
 * @param {Container} container (i.e. the Stage)
 * @returns {Promise<void>}
 */
export async function render(element, container, callback) {
  invariant(container instanceof Container, 'Invalid argument `container`, expected instance of `PIXI.Container`.')

  let root = roots.get(container)
  if (!root) {
    // get the flushed fiber container
    root = PixiFiber.createContainer(container)
    roots.set(container, root)
  }

  // schedules a top level update
  PixiFiber.updateContainer(element, node, null, callback)

  // inject into react devtools
  PixiFiber.injectIntoDevTools({
    bundleType: process.env.NODE_ENV === 'development' ? 1 : 0,
    version: VERSION,
    rendererPackageName: PACKAGE_NAME,
    findHostInstanceByFiber: PixiFiber.findHostInstance,
  })

  // parse input and returns output here ?

  // return the root instance
  return PixiFiber.getPublicRootInstance(node)
}
