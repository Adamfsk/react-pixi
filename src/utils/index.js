import { createElement } from './createElement'
import getTextureFromProps from './getTextureFromProps'

export { createElement, getTextureFromProps }
