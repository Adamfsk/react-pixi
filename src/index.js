import { TYPES } from './utils/element'
import { render } from './render'
import Stage from './stage'
import Provider from './stage/provider'

/**
 * -------------------------------------------
 * Public API
 * -------------------------------------------
 */

export { render, Stage, Provider }

export const BitmapText = TYPES.BitmapText
export const Container = TYPES.Container
export const Graphics = TYPES.Graphics
export const NineSlicePlane = TYPES.NineSlicePlane
export const ParticleContainer = TYPES.ParticleContainer
export const Sprite = TYPES.Sprite
export const Text = TYPES.Text
export const TilingSprite = TYPES.TilingSprite
export const Shape = TYPES.Shape
export const Mesh = TYPES.Mesh
export const Rope = TYPES.Rope
