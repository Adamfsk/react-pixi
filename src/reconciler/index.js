import Reconciler from 'react-reconciler'
import emptyObject from 'fbjs/lib/emptyObject'
import { createElement } from '../utils/createElement'
import now from 'performance-now'

import pkg from '../../package.json'

/**
 * The React Reconciler handling PIXI elements
 *
 * @type {Object}
 */
const PixiFiber = Reconciler({
  appendInitialChild(parent, child) {
    if (parent.appendChild) {
      parent.appendChild(child)
    } else {
      parent.document = child
    }
  },

  createInstance(type, props, internalInstanceHandler) {
    return createElement(type, props, internalInstanceHandler)
  },

  createTextInstance(text, rootContainerInstance, internalInstanceHandler) {
    return text
  },

  finalizeInitialChildren(wordElement, type, props) {
    return false
  },

  getPublicInstance(ins) {
    return ins
  },

  prepareForCommit() {
    // noop
  },

  prepareUpdate(wordElement, type, oldProps, newProps) {
    return true
  },

  resetAfterCommit() {
    // noop
  },

  resetTextContent(wordElement) {
    // noop
  },

  getRootHostContext(rootInstance) {
    // You can use this 'rootInstance' to pass data from the roots.
  },

  getChildHostContext() {
    return emptyObject
  },

  shouldSetTextContent(type, props) {
    return false
  },

  now,

  useSyncScheduling: true,

  mutation: {
    appendChild(parent, child) {
      if (parent.appendChild) {
        parent.appendChild(child)
      } else {
        parent.document = child
      }
    },

    appendChildToContainer(parent, child) {
      if (parent.appendChild) {
        parent.appendChild(child)
      } else {
        parent.document = child
      }
    },

    removeChild(parent, child) {
      parent.removeChild(child)
    },

    removeChildFromContainer(parent, child) {
      parent.removeChild(child)
    },

    inserBefore(parent, child, beforeChild) {
      // noop
    },

    commitUpdate(instance, updatePayload, type, oldProps, newProps) {
      // noop
    },

    commitMount(instance, updatePayload, type, oldProps, newProps) {
      // noop
    },

    commitTextUpdate(textInstance, oldText, newText) {
      textInstance.children = newText
    },
  },
})

export default PixiFiber

export const VERSION = pkg.version

export const PACKAGE_NAME = pkg.name
