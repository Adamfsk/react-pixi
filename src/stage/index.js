import React from 'react'
import PropTypes from 'prop-types'
import invariant from 'fbjs/lib/invariant'
import { PROPS_DISPLAY_OBJECT } from '../utils/props'
import { renderFromComponent } from '../render'

const noop = () => {}

/**
 * -------------------------------------------
 * Stage React Component (use this in react-dom)
 *
 * @usage
 *
 *    const App = () => (
 *      <Stage width={500}
 *             height={500}
 *             options={ backgroundColor: 0xff0000 }
 *             onMount={( renderer, canvas ) => {
 *               console.log('PIXI renderer: ', renderer)
 *               console.log('Canvas element: ', canvas)
 *             }}>
 *    )
 *
 * -------------------------------------------
 */

const propTypes = {
  // dimensions
  width: PropTypes.number,
  height: PropTypes.number,

  // use builtin ticker? default=true
  useTicker: PropTypes.bool,

  // will return canvas and renderer instance
  onMount: PropTypes.func,

  // PIXI options, see http://pixijs.download/dev/docs/PIXI.Application.html
  options: PropTypes.shape({
    antialias: PropTypes.bool,
    autoStart: PropTypes.bool,
    width: PropTypes.number,
    height: PropTypes.number,
    transparent: PropTypes.bool,
    preserveDrawingBuffer: PropTypes.bool,
    resolution: PropTypes.number,
    forceCanvas: PropTypes.bool,
    backgroundColor: PropTypes.number,
    clearBeforeRender: PropTypes.bool,
    roundPixels: PropTypes.bool,
    forceFXAA: PropTypes.bool,
    legacy: PropTypes.bool,
    powerPreference: PropTypes.string,
    sharedTicker: PropTypes.bool,
    sharedLoader: PropTypes.bool,

    // view is optional, use if provided
    view: (props, propName, componentName) => {
      const el = props[propName]
      if (el === undefined) {
        return
      }
      invariant(
        el instanceof HTMLCanvasElement,
        `Invalid prop \`view\` of type ${typeof el}, supplied to ${componentName}, expected \`<canvas> Element\``
      )
    },
  }),
}

const defaultProps = {
  width: 800,
  height: 600,
  useTicker: true,
  onMount: noop,
}

function getCanvasProps(props) {
  const reserved = [...Object.keys(propTypes), ...Object.keys(PROPS_DISPLAY_OBJECT)]

  return Object.keys(props)
    .filter(p => !reserved.includes(p))
    .reduce((all, prop) => ({ ...all, [prop]: props[prop] }), {})
}

class Stage extends React.Component {
  _canvas = null

  componentWillMount() {
    invariant(!!window, `Cannot mount Stage, window object is not defined`)
  }

  componentDidMount() {
    const { children, width, height, options } = this.props

    this.app = new PIXI.Application(width, height, {
      ...options,
      view: this._canvas,
    })

    this.mountNode = renderFromComponent(children, this.app.stage, this, true)
  }

  componentDidUpdate(prevProps, prevState, prevContext) {
    const { children, width, height } = this.props

    // handle resize
    if (prevProps.height !== height || prevProps.width !== width) {
      this.app.resize(width, height)
    }

    // handle resolution ?

    renderFromComponent(children, this.mountNode, this)
  }

  componentWillUnmount() {
    renderFromComponent(null, this.mountNode, this)
  }

  render() {
    const { options } = this.props

    if (!options && options.view) {
      return null
    }

    return <canvas {...getCanvasProps(this.props)} ref={c => (this._canvas = c)} />
  }
}

Stage.propTypes = propTypes
Stage.defaultProps = defaultProps

export default Stage
