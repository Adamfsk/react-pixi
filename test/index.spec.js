import * as index from '../src'

describe('index', () => {
  test('export modules', () => {
    expect(index).toMatchSnapshot()
  })
})
