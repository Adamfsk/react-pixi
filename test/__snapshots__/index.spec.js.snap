// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index export modules 1`] = `
Object {
  "BitmapText": "BitmapText",
  "Container": "Container",
  "Graphics": "Graphics",
  "Mesh": "Mesh",
  "NineSlicePlane": "NineSlicePlane",
  "ParticleContainer": "ParticleContainer",
  "Provider": [Function],
  "Rope": "Rope",
  "Shape": "Shape",
  "Sprite": "Sprite",
  "Stage": [Function],
  "Text": "Text",
  "TilingSprite": "TilingSprite",
  "render": [Function],
}
`;
